### PR TITLE
[staging] rust: 1.45 -> 1.46.0

### PR DESCRIPTION
**Pull request by basvandijk**
###### Motivation for this change

This adds the rust toolchain 1.46.0:

https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1460-2020-08-27

Because rustc-1.46.0 enables static PIE by default for `x86_64-unknown-linux-musl` this release will suffer from: https://github.com/NixOS/nixpkgs/issues/94228. So this PR doesn't remove the 1.45.2 release.

This commit also specifies the right LLVM packages to use for each rust release.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by basvandijk** [comment]
@GrahamcOfBorg build fd firefox thunderbird

**Comment by basvandijk** [comment]
Building `fd` on `x86_64-linux` succeeds.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -9518,15 +9518,22 @@ in
 
   rust_1_44 = callPackage ../development/compilers/rust/1_44.nix {
     inherit (darwin.apple_sdk.frameworks) CoreFoundation Security;
+    llvmPackages = llvmPackages_9;
   };
   rust_1_45 = callPackage ../development/compilers/rust/1_45.nix {
```
Comment (kevincox) [inline pkgs/top-level/all-packages.nix:9523]: It might be nice to leave a comment explaining why the older versions are left available.

**Review by kevincox — APPROVED** [review]

**Comment by jonringer** [comment]
@GrahamcOfBorg eval

**Review by lblasc — APPROVED** [review]

**Comment by 06kellyjac** [comment]
Build of deno `1.4.0` on `x86_64-linux` was a success.

**Review by jonringer — COMMENTED** [review]
diff LGTM
commit message should be:
```
rust: 1.45 -> 1.46.0
```
as you're bumping the default

**Comment by 06kellyjac** [comment]
Are these the only 2 things that need resolving to get this merged?:

* [ ] comment explaining why the older versions are left available
* [ ] Update commit message to `rust: 1.45 -> 1.46.0`

**Comment by oxalica** [comment]
I noticed that we still lock `firefox` and `thunderbird` on `rustPackages_1_44`.
Do they actually work on 1.46? I think the `musl` issue mentioned should not block this.
If they are built successfully with 1.46, at least we can remove rust 1.44 from nixpkgs.

**Review by oxalica — CHANGES_REQUESTED** [review]
`firefox`, `thunderbird` and `thunderbird-78` are built successfully with latest `rustPackages_1_46_0`.
We can remove the override (rust 1.44) in `all-packages.nix` at line 20463, 23292 and 23300.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (basvandijk) [reply pkgs/top-level/all-packages.nix:9642]: Done

**Comment by basvandijk** [comment]
@oxalica 
> firefox, thunderbird and thunderbird-78 are built successfully with latest rustPackages_1_46_0.

Thanks for checking!

> We can remove the override (rust 1.44) in all-packages.nix at line 20463, 23292 and 23300.

Done

@06kellyjac

> * [x] comment explaining why the older versions are left available

Done

> * [x] Update commit message to `rust: 1.45 -> 1.46.0`

Done

**Comment by basvandijk** [comment]
@GrahamcOfBorg eval

**Review by oxalica — APPROVED** [review]

**Pull request merged**